doc/source/ray-core/internals/autoscaler-v2.rst

```diff
@@ -0,0 +1,204 @@
+
+
+Overview
+--------
+
+The autoscaler is responsible for resizing the cluster based on resource demand including tasks, actors, and placement groups.
+To achieve this, it follows a structured process: evaluating worker group configurations, periodically reconciling cluster state with user constraints, applying bin-packing strategies to pending workload demands, and interacting with cloud providers through the Instance Manager.
+The following sections describe these components in detail.
+
+Worker Group Configurations
+---------------------------
+
+Worker groups (also referred to as node types) define the sets of nodes that the Ray autoscaler scales.
+Each worker group represents a logical category of nodes with the same resource configurations, such as CPU, memory, GPU, or custom resources.
+
+The autoscaler dynamically adjusts the cluster size by adding or removing nodes within each group as workload demands change. In other words, it scales the cluster by modifying the number of nodes per worker group according to the specified scaling rules and resource requirements.
+
+Worker groups can be configured in these ways:
+
+- The `available_node_types <https://docs.ray.io/en/latest/cluster/vms/references/ray-cluster-configuration.html#node-types>`__ field in the Cluster YAML file, if you are using the ``ray up`` cluster launcher.
+- The `workerGroupSpecs <https://docs.ray.io/en/latest/cluster/kubernetes/user-guides/config.html#pod-configuration-headgroupspec-and-workergroupspecs>`__ field in the RayCluster CRD, if you are using KubeRay.
+
+The configuration specifies the logical resources each node has in a worker group, along with the minimum and maximum number of nodes that should exist in each group.
+
+.. note::
+
+   Although the autoscaler fulfills pending resource demands and releases idle nodes, it doesn't perform the actual scheduling of Ray tasks, actors, or placement groups. Scheduling is handled internally by Ray.
+   The autoscaler does its own simulation of scheduling decisions on pending demands periodically to determine which nodes to launch or to stop. See the next sections for details.
+
+
+Periodic Reconciliation
+-----------------------
+
+The entry point of the autoscaler is `monitor.py <https://github.com/ray-project/ray/blob/03491225d59a1ffde99c3628969ccf456be13efd/python/ray/autoscaler/v2/monitor.py#L332>`__, which starts a GCS client and runs the reconciliation loop.
+
+This process is launched on the head node by the `start_head_processes <https://github.com/ray-project/ray/blob/03491225d59a1ffde99c3628969ccf456be13efd/python/ray/_private/node.py#L1439>`__ function when using the ``ray up`` cluster launcher.
+When running under KubeRay, it instead runs as a separate autoscaler container in the Head Pod.
+
+.. note::
+
+   In the case of the cluster launcher, if the autoscaler process crashes, then there is no autoscaling.
+   While in the case of KubeRay, Kubernetes restarts the autoscaler container if it crashes.
```

Comment:
link to the restart code? In general, try to have many code links if possible since the audience is developers.

Reply:
Updated the statement to indicate that it uses the default Kubernetes behavior, so there is no explicit code doing the restart.